### PR TITLE
Connect to efs from bastion

### PR DESCRIPTION
**Pull request by MancunianSam**
- Add parameters to the Jenkins job to optionally connect to database/efs
- Change Jenkins job to avoid running the create user script if the user was never created.
- Conditionally add policies to the bastion role to allow it to connect to efs/database
- Add in extra variables for the EC2 role and user data.
- Optionally add in security group rules to allow the bastion to connect.
- Upgrade AWS provider as the older version didn't allow filtering file systems by tag name.

### CODE REVIEW

File: terraform/bastion/Jenkinsfile [hunk 1]
```diff
@@ -43,7 +52,7 @@ pipeline {
     stage('Create user if creating bastion') {
```
Comment (TomJKing) [inline terraform/bastion/Jenkinsfile:52]: Maybe change the `stage` description as this only applies if connecting to the DB, not the other connections: "Create user if connecting to database"

File: terraform/bastion/Jenkinsfile [hunk 2]
```diff
@@ -75,14 +84,17 @@ pipeline {
...
+        TF_VAR_connect_to_export_efs="${params.CONNECT_TO_EXPORT_EFS}"
       }
       stages {
         stage('Set up Terraform workspace') {
           steps {
             script {
               dir("terraform/bastion") {
                 echo 'Initializing Terraform...'
-                sh "git clone https://github.com/nationalarchives/tdr-terraform-modules.git"
+                sh "git clone --branch connect-to-efs-from-bastion https://github.com/nationalarchives/tdr-terraform-modules.git"
```
Comment (TomJKing) [inline terraform/bastion/Jenkinsfile:97]: Should the `--branch` param be in the command?

File: terraform/bastion/Jenkinsfile [hunk 3]
```diff
@@ -9,6 +9,9 @@ pipeline {
     choice(name: "STAGE", choices: ["intg", "staging", "prod"], description: "The stage you are creating the instance in")
     choice(name: "COMMAND", choices: ["apply", "destroy"], description: "Apply will create the instance")
     text(name: "PUBLIC_KEY", description: "Public key. This is not needed if you're not going to use ssh tunnelling")
+    booleanParam(name: 'CONNECT_TO_BACKEND_CHECKS_EFS', defaultValue: false, description: 'Allow the bastion to connect to the backend checks EFS volume')
```
Comment (TomJKing) [inline terraform/bastion/Jenkinsfile:12]: Does one of these have to be set to `true` for the job to work? What happens if more than one is set to `true`?

**Review by TomJKing — COMMENTED** [review]
Looks good. Some minor comments.

File: terraform/bastion/Jenkinsfile (hunk 1 above)
Comment (MancunianSam) [reply terraform/bastion/Jenkinsfile:52]: Makes sense, I've changed it.

File: terraform/bastion/Jenkinsfile (hunk 2 above)
Comment (MancunianSam) [reply terraform/bastion/Jenkinsfile:97]: Nope, that was just for testing. I've taken it out.

File: terraform/bastion/Jenkinsfile (hunk 3 above)
Comment (MancunianSam) [reply terraform/bastion/Jenkinsfile:12]: They can all be set to false but you'll be left with a bastion that can't connect to anything. If you select more of them, you get access to whichever ones you've checked.

Comment (TomJKing) [reply terraform/bastion/Jenkinsfile:12]: Ah, that makes sense.

**Review by TomJKing — APPROVED** [review]

**Pull request merged**